internals/osutil/mkdirallchown.go

```diff
@@ -84,18 +91,30 @@ func mkdirChown(path string, perm os.FileMode, uid sys.UserID, gid sys.GroupID)
 	if err != nil {
 		return err
 	}
+	if flags&AtomicWriteChmod != 0 {
```

Comment:
I think this block should be after the `defer fd.Close()`, otherwise we'll leave this open if the `fd.Chmod` returns an error.